compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -755,8 +755,19 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     self.fail(location, format!("explicit `{:?}` is forbidden", kind));
                 }
             }
-            StatementKind::StorageLive(..)
-            | StatementKind::StorageDead(..)
+            StatementKind::StorageLive(local) => {
+                if self.reachable_blocks.contains(location.block) {
```

Comment:
Can you reference the corresponding comment from visit_local method here? (Conceptually this shouldn't be part of a validator, but rather a separate optional lint pass, but as long as that comment is true that seems reasonable).